rest/blip_sync.go

```diff
@@ -932,30 +936,46 @@ func (bh *blipHandler) handleRev(rq *blip.Message) error {
 			deltaSrcBody[db.BodyAttachments] = map[string]interface{}(deltaSrcRev.Attachments)
 		}
 
-		deltaSrcMap := map[string]interface{}(deltaSrcBody)
-		err = base.Patch(&deltaSrcMap, newDoc.Body())
+		var body db.Body
+
+		// Unmarshal delta including properties for patching
+		err = rq.ReadJSONBody(&body)
+		if err != nil {
+			return base.HTTPErrorf(http.StatusBadRequest, "error unmarshalling body: %v", err)
+		}
+
+		patchedBody := map[string]interface{}(deltaSrcBody)
+
+		err = base.Patch(&patchedBody, body)
 		if err != nil {
 			return base.HTTPErrorf(http.StatusInternalServerError, "Error patching deltaSrc with delta: %s", err)
 		}
 
-		newDoc.UpdateBody(deltaSrcMap)
-		bh.Logf(base.LevelTrace, base.KeySync, "docID: %s - body after patching: %v", base.UD(docID), base.UD(deltaSrcMap))
+		bh.Logf(base.LevelTrace, base.KeySync, "docID: %s - body after patching: %v", base.UD(docID), base.UD(patchedBody))
 		bh.db.DbStats.StatsDeltaSync().Add(base.StatKeyDeltaPushDocCount, 1)
+
+		newBody = patchedBody
 	}
 
-	// Handle and pull out expiry
-	if bytes.Contains(bodyBytes, []byte(db.BodyExpiry)) {
-		body := newDoc.Body()
-		expiry, err := body.ExtractExpiry()
+	if hasAttachments || hasExpiry {
+		if newBody == nil {
+			newBody, err = newDoc.GetBody()
+			if err != nil {
+				return base.HTTPErrorf(http.StatusInternalServerError, "error getting IncomingDocument body: %v", err)
+			}
+		}
+
+		newDoc, err = newBody.ToIncomingDocument()
 		if err != nil {
-			return base.HTTPErrorf(http.StatusBadRequest, "Invalid expiry: %v", err)
+			return base.HTTPErrorf(http.StatusInternalServerError, "error occurred building IncomingDocument from body: %v", err)
+		}
+	} else if newBody != nil {
+		err := newDoc.UpdateBody(newBody)
+		if err != nil {
+			return base.HTTPErrorf(http.StatusInternalServerError, "error occurred updating IncomingDocument body: %v", err)
 		}
-		newDoc.DocExpiry = expiry
-		newDoc.UpdateBody(body)
 	}
```

Comment:
We need more comments around here since we added the "newBody" for an optimisation. It's confusing to follow without prior knowledge of what it is.

E.g. justify why the `else if` block is here (detail the att+delta scenario)

Reply:
Written some comments that I think should help the explanation.